File: regalia-logansquare/src/main/java/com/commit451/regalia/logansquare/RealmListLoganSquareConverter.java
package com.commit451.regalia.logansquare;

import com.bluelinelabs.logansquare.JsonMapper;
import com.bluelinelabs.logansquare.LoganSquare;
import com.bluelinelabs.logansquare.NoSuchTypeConverterException;
import com.bluelinelabs.logansquare.typeconverters.TypeConverter;
import com.fasterxml.jackson.core.JsonGenerator;
import com.fasterxml.jackson.core.JsonParser;
import com.fasterxml.jackson.core.JsonToken;

import java.io.IOException;

import io.realm.RealmList;
import io.realm.RealmObject;

/**
 * The universal converter for RealmList.
 * You can use this with RealmObjects:
 * 1. Annotated with @JsonObject
 * 2. With custom converters.
 *
 * @author Krzysztof Miemiec
 */
public abstract class RealmListLoganSquareConverter<T extends RealmObject> implements TypeConverter<RealmList<T>> {

    private Class<T> clazz;
    private TypeConverter<T> typeConverter;
    private JsonMapper<T> mapper;

    public RealmListLoganSquareConverter(Class<T> clazz) {
        this.clazz = clazz;
    }

    private void init() {
        if (typeConverter == null && mapper == null) {
            try {
                typeConverter = LoganSquare.typeConverterFor(clazz);
            } catch (NoSuchTypeConverterException e) {
                mapper = LoganSquare.mapperFor(clazz);
            }
        }
    }

    @Override
    public RealmList<T> parse(JsonParser jsonParser) throws IOException {
        init();
        RealmList<T> list = new RealmList<>();
        if (jsonParser.getCurrentToken() == JsonToken.START_ARRAY) {
            while (jsonParser.nextToken() != JsonToken.END_ARRAY) {
                T object = typeConverter != null ? typeConverter.parse(jsonParser) : mapper.parse(jsonParser);
                if (object != null) {
                    list.add(object);
                }
            }
        }
        return list;
    }

    @Override
    public void serialize(RealmList<T> list, String fieldName, boolean writeFieldNameForObject, JsonGenerator jsonGenerator) throws IOException {
        if (writeFieldNameForObject) {
            if (list == null) {
                return;
            }
            jsonGenerator.writeFieldName(fieldName);
        }
        if (list == null) {
            jsonGenerator.writeNull();
            return;
        }
        jsonGenerator.writeStartArray();

        init();
        for (T object : list) {
            if (object != null) {
                if (typeConverter != null) {
                    typeConverter.serialize(object, fieldName, false, jsonGenerator);
                } else {
                    mapper.serialize(object, jsonGenerator, true);
                }
            } else {
                jsonGenerator.writeNull();
            }
        }
        jsonGenerator.writeEndArray();
    }
}

File: regalia-logansquare/src/main/java/com/commit451/regalia/logansquare/RealmListLoganSquareConverter.kt
package com.commit451.regalia.logansquare

import com.bluelinelabs.logansquare.JsonMapper
import com.bluelinelabs.logansquare.LoganSquare
import com.bluelinelabs.logansquare.NoSuchTypeConverterException
import com.bluelinelabs.logansquare.typeconverters.TypeConverter
import com.fasterxml.jackson.core.JsonGenerator
import com.fasterxml.jackson.core.JsonParser
import com.fasterxml.jackson.core.JsonToken
import io.realm.RealmList
import io.realm.RealmObject
import java.io.IOException

/**
 * The universal converter for RealmList.
 * You can use this with RealmObjects:
 * 1. Annotated with @JsonObject
 * 2. With custom converters.
 */
abstract class RealmListLoganSquareConverter<T : RealmObject?> : TypeConverter<RealmList<T>?> {
    private lateinit var typeConverter: TypeConverter<T>
    private lateinit var mapper: JsonMapper<T>
    abstract val objectClass: Class<T>?
    private fun init() {
        try {
            typeConverter = LoganSquare.typeConverterFor(objectClass)
        } catch (e: NoSuchTypeConverterException) {
            mapper = LoganSquare.mapperFor(objectClass)
        }
    }

    @Throws(IOException::class)
    override fun parse(jsonParser: JsonParser): RealmList<T>? {
        init()
        val list = RealmList<T>()
        if (jsonParser.currentToken == JsonToken.START_ARRAY) {
            while (jsonParser.nextToken() != JsonToken.END_ARRAY) {
                val element = typeConverter.parse(jsonParser)
                if (element != null) {
                    list.add(element)
                }
            }
        }
        return list
    }

    @Throws(IOException::class)
    override fun serialize(list: RealmList<T>?, fieldName: String, writeFieldNameForObject: Boolean, jsonGenerator: JsonGenerator) {
        if (writeFieldNameForObject) {
            if (list == null) {
                return
            }
            jsonGenerator.writeFieldName(fieldName)
        }
        if (list == null) {
            jsonGenerator.writeNull()
            return
        }
        jsonGenerator.writeStartArray()
        init()
        for (t in list) {
            if (t != null) {
                typeConverter.serialize(t, fieldName, false, jsonGenerator)
            } else {
                jsonGenerator.writeNull()
            }
        }
        jsonGenerator.writeEndArray()
    }
}